File: features/list/src/test/java/com/vp/list/viewmodel/ListViewModelTest.java
package com.vp.list.viewmodel;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Observer;

import com.vp.list.model.ListItem;
import com.vp.list.model.SearchResponse;
import com.vp.list.service.SearchService;

import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.internal.verification.AtLeast;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import retrofit2.mock.Calls;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ListViewModelTest {
    @Rule
    public InstantTaskExecutorRule instantTaskRule = new InstantTaskExecutorRule();

    @Test
    public void shouldReturnErrorState() {
        //given
        SearchService searchService = mock(SearchService.class);
        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.failure(new IOException()));
        ListViewModel listViewModel = new ListViewModel(searchService);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        assertThat(listViewModel.observeMovies().getValue().getListState()).isEqualTo(ListState.ERROR);
    }

    @Test
    public void shouldReturnInProgressState() {
        //given
        SearchService searchService = mock(SearchService.class);
        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.response(mock(SearchResponse.class)));
        ListViewModel listViewModel = new ListViewModel(searchService);
        Observer<SearchResult> mockObserver = (Observer<SearchResult>) mock(Observer.class);
        listViewModel.observeMovies().observeForever(mockObserver);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        verify(mockObserver).onChanged(SearchResult.inProgress());
    }

    @Test
    public void shouldReturnSuccessState() {

        //given
        SearchService searchService = mock(SearchService.class);
        SearchResponse searchResponse = mock(SearchResponse.class);
        List<ListItem> listItems = new ArrayList<>();
        listItems.add(mock(ListItem.class));
        listItems.add(mock(ListItem.class));

        when(searchService.search(anyString(), anyInt())).thenReturn(Calls.response(searchResponse));
        when(searchResponse.getSearch()).thenReturn(listItems);
        when(searchResponse.getTotalResults()).thenReturn(listItems.size());

        ListViewModel listViewModel = new ListViewModel(searchService);
        Observer<SearchResult> mockObserver = (Observer<SearchResult>) mock(Observer.class);
        ArgumentCaptor<SearchResult> captor = ArgumentCaptor.forClass(SearchResult.class);
        listViewModel.observeMovies().observeForever(mockObserver);

        //when
        listViewModel.searchMoviesByTitle("title", 1);

        //then
        verify(mockObserver, new AtLeast(1)).onChanged(captor.capture());

        Assert.assertEquals(SearchResult.inProgress(), captor.getAllValues().get(0));
        Assert.assertEquals(SearchResult.success(listItems, listItems.size()), captor.getAllValues().get(1));
    }

}

File: features/list/src/test/java/com/vp/list/viewmodel/ListViewModelTest.kt
package com.vp.list.viewmodel

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.Observer
import com.vp.list.model.ListItem
import com.vp.list.model.SearchResponse
import com.vp.list.service.SearchService
import com.vp.list.viewmodel.ListState.ERROR
import com.vp.list.viewmodel.SearchResult.Companion.inProgress
import com.vp.list.viewmodel.SearchResult.Companion.success
import org.assertj.core.api.Assertions
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.mockito.internal.verification.AtLeast
import retrofit2.mock.Calls
import java.io.IOException

class ListViewModelTest {
    @get:Rule
    var instantTaskRule = InstantTaskExecutorRule()

    @Test
    fun shouldReturnErrorState() { //given
        val searchService = Mockito.mock(SearchService::class.java)
        Mockito.`when`(
            searchService.search(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyInt()
            )
        ).thenReturn(Calls.failure(IOException()))
        val listViewModel = ListViewModel(searchService)
        //when
        listViewModel.searchMoviesByTitle("title", 1)
        //then
        Assertions.assertThat(listViewModel.observeMovies().value?.listState)
            .isEqualTo(ERROR)
    }

    @Test
    fun shouldReturnInProgressState() { //given
        val searchService = Mockito.mock(SearchService::class.java)
        val searchResponse: SearchResponse = SearchResponse(" True", emptyList())
        Mockito.`when`(
            searchService.search(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyInt()
            )
        ).thenReturn(
            Calls.response(
                searchResponse
            )
        )
        val listViewModel = ListViewModel(searchService)
        val mockObserver =
            Mockito.mock(Observer::class.java) as Observer<SearchResult>
        listViewModel.observeMovies().observeForever(mockObserver)
        //when
        listViewModel.searchMoviesByTitle("title", 1)
        //then
        Mockito.verify(mockObserver)
            .onChanged(inProgress())
    }

    @Test
    fun shouldReturnSuccessState() { //given
        val searchService = Mockito.mock(SearchService::class.java)
        val listItem = ListItem("mock", "mock", "mock", "mock")
        val listItems: MutableList<ListItem> = ArrayList()
        listItems.add(listItem.copy())
        listItems.add(listItem.copy())
        val searchResponseMock = SearchResponse("True", listItems)

        Mockito.`when`(
            searchService.search(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyInt()
            )
        ).thenReturn(Calls.response(searchResponseMock))

        val listViewModel = ListViewModel(searchService)
        val mockObserver =
            Mockito.mock(Observer::class.java) as Observer<SearchResult>
        val captor =
            ArgumentCaptor.forClass(
                SearchResult::class.java
            )
        listViewModel.observeMovies().observeForever(mockObserver)

        listViewModel.searchMoviesByTitle("title", 1)

        Mockito.verify(mockObserver, AtLeast(1)).onChanged(captor.capture())
        Assert.assertEquals(inProgress(), captor.allValues[0])
        Assert.assertEquals(
            success(listItems, listItems.size),
            captor.allValues[1]
        )
    }
}